### ISSUE TITLE
Profiling report with branch coverage for templates is misleading

### ISSUE THREAD

**Issue by stma247** [llvm/llvm-project#111743]
The export mode of tool "llvm-cov" for format "lcov" does not combine the branch counts of C++ template functions instantiations.

A preceding discussion describes the issue in detail: 
https://discourse.llvm.org/t/profiling-report-with-branch-coverage-for-templates-is-misleading/82323

**Comment by stma247** [comment]
I am already working on a solution. The PR will follow soon.

**Comment by chapuni** [comment]
Thanks for working on it.
Could it be applied to text and html as well?

**Comment by evodius96** [comment]
> Thanks for working on it. Could it be applied to text and html as well?

This is only really for LCOV exporting.  The text and html from `llvm-cov show` summaries show the right information.

**Comment by chapuni** [comment]
From llvm/test/tools/llvm-cov/branch-templates.cpp
```
   12|       |template<typename T>
   13|      3|int func(T x) {
   14|      3|  if(x)       // CHECK: |  Branch ([[@LINE]]:6): [True: 0, False: 1]
  ------------------
  |  Branch (14:6): [True: 0, False: 1]
  |  Branch (14:6): [True: 1, False: 0]
  |  Branch (14:6): [True: 0, False: 1]
  ------------------
   15|      1|    return 0; // CHECK: |  Branch ([[@LINE-1]]:6): [True: 1, False: 0]
   16|      2|  else        // CHECK: |  Branch ([[@LINE-2]]:6): [True: 0, False: 1]
   17|      2|    return 1;
   18|      0|  int j = 1;
   19|      0|}
```

I'll investigate further later.

**Comment by evodius96** [comment]
Hi @chapuni  

Sorry, I didn't elaborate -- this is primarily an effect of the summary report.  With `llvm-cov show` the summary report correctly merges the data across the corresponding instantiations showing the union of the data, whereas with `llvm-cov export --format=lcov` it was not correctly rendering the union with BRDA.  As a result, LCOV was not displaying the summary correctly.

The source-level report that you show shows the individual branch coverage information across the instantiations, but the report summary for this should have merged this data and ought to show 100% coverage for the condition.

**Comment by chapuni** [comment]
@evodius96 I supposed this issues as more generic since @stma247's branch achieves partially what I wanted.

```
// UNIFY-DAG: BRDA:14,0,0,1
// UNIFY-DAG: BRDA:14,0,1,2
```

May I file another issue for merging Branches in source view? I think it's odd that merging is done in Line counts but not in Branches. Individual Branches can be seen in expanded view.

**Comment by evodius96** [comment]
> May I file another issue for merging Branches in source view? I think it's odd that merging is done in Line counts but not in Branches. Individual Branches can be seen in instantiations view.

That's true. You could merge the branches in the common view as long as you can see each instantiation.

**Comment by bartlettroscoe** [comment]
> > Thanks for working on it. Could it be applied to text and html as well?
> 
> This is only really for LCOV exporting. The text and html from `llvm-cov show` summaries show the right information.

Yes, but I will argue you still want to have the overall coverage metrics for the template function, in addition to its specific instantiations.  The way `llvm-cov show`, for example, displays coverage information for template functions with lots of instantiations is about worthless (you can't see the forest for the trees).  I will post an issue about how to fix this shortly (i.e. wrapping details for the instantiations in `<details> ... </details>` blocks).  And for that, we will need combined branch metrics for the template function across all instantiations.
